Flag for tracking when Welcome Guid (NUX modal or Tour) is opened via the More Menu

autumnfjeld opened the pull request:
## Changes proposed in this Pull Request

* Add `isTourManuallyOpened` and `isGuideManuallyOpened` flag to indicate in tracking events if user is seeing Welcome Guide (Tour or NUX) from More Menu OR if user is a new user, viewing the Welcome Guide automatically, when the open the editor for the first time.

Want to note that I've found the `isNuxEnabled` language confusing because it is toggled on and off when user accesses the Guide from the more menu.   Logic would be easier to follow using something like Gutenberg's  [toggleFeature( 'welcomeGuide' )](https://github.com/WordPress/gutenberg/blob/c88866cd91ea3eb7990a68978e03e2366ed7106c/packages/edit-post/src/components/welcome-guide/index.js#L38) and from the [MoreMenu](https://github.com/WordPress/gutenberg/blob/c88866cd91ea3eb7990a68978e03e2366ed7106c/packages/edit-post/src/plugins/welcome-guide-menu-item/index.js#L17)
Putting on my todo list for post MVP work.

## Testing instructions

If testing on local standalone WP:
* Pull down this branch
* Add define( 'SHOW_WELCOME_TOUR', true ); to wp-config.php

If testing on sandbox:
* Pull down this branch and sync to sandbox `yarn dev --sync`
* Add define( 'SHOW_WELCOME_TOUR', true ); to your /home/wpcom/public_html/wp-content/mu-plugins/0-sandbox.php file. This will force the Welcome Tour to show every time the editor loads.

❗ In dev tools Application tab you must clear the site's local storage to ensure the @automattic/nux store is reset. Must be done in order to show the Welcome Guide on first load.

### #1 Test Welcome Tour

* Open a test site 
* In dev tools Application tab: Clear local storage for the site
* In dev tools Network tab: and filter by `calypso` and keep tab open to watch network requests
* Open the editor , and view the network requests and look for  the tracking event
  * `calypso_editor_wpcom_tour_open`
    - data: {is_gutenboarding:}
    - data: {is_manually_opened: **false**}  🔫
![Screen Shot 2021-01-14 at 14 05 30](https://user-images.githubusercontent.com/5665959/104539420-9968af80-5671-11eb-99d1-ea13b11f373d.png)
* Close the Tour
* Got to more menu and open "Welcome Guide"
* Observe same open event in dev tools, but now {is_manually_opened: **true**} 🔫


### #2 Test NUX modal
* Do a little hack to get NUX modal to show instead of the tour:  make[ `load_wpcom_block_editor_nux() `](https://github.com/Automattic/wp-calypso/blob/12f778c15658ed71646145aab9f83586900d04e4/apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php#L243) require the NUX modal and NOT the tour
* Repeat test steps given above but now for the NUX modal


<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->


Fixes #

autumnfjeld committed:
Idea for tracking when NUX modal or Tour is opened via the More Menu. Not working code. Gathering feedback for idea

autumnfjeld commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/disable-core-nux.js:
```diff
@@ -22,5 +22,6 @@ subscribe( () => {
 		dispatch( 'automattic/nux' ).setWpcomNuxStatus( { isNuxEnabled: true } );
+		dispatch( 'automattic/nux' ).toggleWpcomTourManualOpenStatus( { isTourManuallyOpen: true } );
 	}
 } );
```
My assessment of the above lines of code.  These are triggered when user selects "Welcome Guide" from MoreMenu
22) check if the welcomeGuide has been activated in core
23) toggle the feature in core (presumably turning it off) since it was just activated
24) set a new property that explicitly flags when the user manually opens the Tour 

We need a way to differentiate the Tour (or NUX modal) being opened automatically (the first time user opens editor) vs when the user opens it from the More Menu.

A better refactor would be to contain the usage of the `isNuxEnabled` purely for the first time viewing of the Tour (or NUX modal)  new-user-experience, thus respecting the '`nux`' meaning.   I'm not sure if it is wise to do that refactor now, or wait until after A/B testing.  My gut feeling is leave a comment and tidy up after A/B testing.

autumnfjeld committed:
To allow tracking of WelcomeGuide (NUX or Tour) as first-editor-view or from more menu, add isTourManuallyOpened to automattic/nux store

autumnfjeld committed:
Differentiate tracking of WelcomeGuide (NUX) open event as first-editor-view vs from more menu by adding isGuideManuallyOpened to automattic/nux store

autumnfjeld committed:
Tour: set isTourManuallyOpened to false on dismiss

autumnfjeld committed:
Tour Content: add commas to two slide sentances

autumnfjeld committed:
Add comment for areTipsEnabled and do not set manual flag in this conditional

autumnfjeld committed:
Add comment to improve variable naming to differentiate from user setting nux and feature active status

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-launch.js:
```diff
@@ -53,6 +56,7 @@ function LaunchWpcomWelcomeTour() {
 
 		recordTracksEvent( 'calypso_editor_wpcom_tour_open', {
 			is_gutenboarding: window.calypsoifyGutenberg?.isGutenboarding,
+			is_manually_opened: isTourManuallyOpened,
```
`isTourManuallyOpened` needs to be added to the list of `useEffect` dependencies.

autumnfjeld replied:
Ah great catch!!!   I was observing some quirky behavior with this earlier but I couldn't consistently reproduce it, I think the missing dependency in useEffect was the root cause.

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/disable-core-nux.js:
```diff
@@ -21,6 +21,9 @@ subscribe( () => {
+		dispatch( 'automattic/nux' ).setWpcomNuxStatus( {
+			isNuxEnabled: true,
+		} );
+		dispatch( 'automattic/nux' ).setTourOpenStatus( { isTourManuallyOpened: true } );
```
Rather than dispatching to actions to the same store maybe it makes sense to coming them?

```typescript
 		dispatch( 'automattic/nux' ).setWpcomNuxStatus( {
 			isNuxEnabled: true,
 			tourWasOpenedManually: true,
 		} );
```

It made me think though, is there ever a time when `isNuxEnabled` is set to `true` and it wasn't a manual opening? In which case why's the extra flag needed? 🤔

p-jackson replied:
I must just be confused, because it's working in manual testing

autumnfjeld replied:
> It made me think though, is there ever a time when isNuxEnabled is set to true and it wasn't a manual opening? In which case why's the extra flag needed? 🤔

For a brand new user, `isNuxEnabled = true` and the WelcomeGuide (Tour or NUX) opens automatically when the user opens the editor for the first time.  So yes, that is the time. :) That is the scenario we want to capture the "First time" viewing.  
After user dismisses the Guide, isNuxEnabled is set to false.   
For me, what has been a source of confusion is that this new-user-flag, which is a user setting, is then used a toggle when user accesses the More Menu.  I'd like to clean that logic up and use terminology similar to GB [isFeatureActive](https://github.com/WordPress/gutenberg/blob/c88866cd91ea3eb7990a68978e03e2366ed7106c/packages/edit-post/src/components/welcome-guide/index.js#L22-L23).    This isn't important for MVP though, so going on my list of todos.

autumnfjeld replied:
> Rather than dispatching to actions to the same store maybe it makes sense to combine them?

Ramon suggested the same.  I decided not to do that because of a reason that isn't coming to mind now! Maybe just clarity and because the 'nux' terminology made the logic harder for me to understand.  (Because I think the open / manual toggles are about the state of feature, not the state of the user.)

p-jackson replied:
Yeah "nux" really seems to be a catch all phrase that gets used for so much! 😭

p-jackson reviewed (commented):
Just taking a quick look at the code before testing manually

p-jackson commented:
> #1 Test Welcome Tour

Everything works great

> #2 Test NUX modal

The "modal opened" event doesn't include a `is_manually_open` prop. Is it supposed to?

Edit: I'll mark as approved because the welcome tour event Is working as expected. But my reading of the PR description is that the existing NUX modal's events are supposed to have changed too, and from reading and testing the code that's not happening for me. If I've misunderstood then this is good to go.

autumnfjeld committed:
WelcomeTour launch: add isTourManuallyOpened to list of useEffect dependencies

autumnfjeld commented:
> The "modal opened" event doesn't include a `is_manually_open` prop. Is it supposed to?


It should 100% have a `is_manually_open` introducing that prop in the tracking event is the reason for the PR.

https://github.com/Automattic/wp-calypso/blob/30c6fe2f3691e3b8ee84a7e914a1899f30861eaa/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/wpcom-nux.js#L65

https://github.com/Automattic/wp-calypso/blob/df4d108c41efc7c12f43e74d98989dcd99d11836/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-launch.js#L59

You can see the prop in this screenshot. 
![Screen Shot 2021-01-14 at 17 08 39](https://user-images.githubusercontent.com/5665959/104551536-2c621380-568b-11eb-989f-63697bf7d0da.png)

a8ci18n commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-content.js:
```diff
@@ -34,7 +34,7 @@ function getTourContent() {
 			heading: __( 'Everything is a block', 'full-site-editing' ),
 			description: __(
-				'In the WordPress Editor paragraphs, images, and videos are all blocks.',
+				'In the WordPress Editor, paragraphs, images, and videos are all blocks.',
```
:ok: This change will be queued for retranslation. We'll use the existing translations in the meantime.

a8ci18n commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/tour-content.js:
```diff
@@ -72,7 +72,7 @@ function getTourContent() {
 		{
 			heading: __( 'Drag & drop', 'full-site-editing' ),
-			description: __( 'To move blocks around click and drag the handle.', 'full-site-editing' ),
+			description: __( 'To move blocks around, click and drag the handle.', 'full-site-editing' ),
```
:ok: This change will be queued for retranslation. We'll use the existing translations in the meantime.

autumnfjeld commented:
@p-jackson 
> The "modal opened" event doesn't include a `is_manually_open` prop. Is it supposed to?
> 

Ah! The prop is `is_manually_opened` with ed.   Did you observe the event with `is_manually_opened`?

ramonjd commented:
Also works great for me!

Here are the params for the tracking GIF when the tour opened automatically (I trigged it by clearing my local storage hence the reason why is_gutenboarding === false:

<img width="485" alt="Screen Shot 2021-01-14 at 8 00 07 pm" src="https://user-images.githubusercontent.com/6458278/104568116-c2a23380-56a3-11eb-902d-455ba9b9eb7e.png">


And clicking on the link in the menu:

<img width="400" alt="Screen Shot 2021-01-14 at 8 00 47 pm" src="https://user-images.githubusercontent.com/6458278/104568109-c03fd980-56a3-11eb-83c0-ca6bd09d3903.png">

roo2 commented:
I tested the welcome tour and the older nux modal, I found a few things not quite as I expected with the nux modal.

Also I wasn't certain the best way to test the older nux modal, it makes more sense to me to simply disable the SHOW_WELCOME_TOUR flag, would't that be more correct then editing the `load_wpcom_block_editor_nux` function? (I'm getting different results)

I also observed the `is_manually_opened` flag being set as expected 
* true when opened by the menu link ✅ 
* false when opened automatically for a user with cleared local storage ✅ 

For the old nux modal, I tried to test it by disabling the `SHOW_WELCOME_TOUR` flag in `sandbox.php`
* I didn't see an `is_manually_opened` flag being passed along to the tracking endpoint 🔴 

I also tried to show the old nux modal my editing the `load_wpcom_block_editor_nux` function:
* When I deleted the local storage and reloaded the page, `is_manually_opened` was not send to the tracking endpoint❓ 
** It should probably send `is_manually_opened: false` the same as the tour
* I dismissed the modal and reloaded the page, the modal did not show ✅ 
* I opened the guide through the menu link and an event was fired with `is_manually_opened: true`
* I reloaded the page and the nux modal was shown again with `is_manually_opened:true` ❓ 
** I think it should be set to false the second time? or just not open?

p-jackson commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/wpcom-nux.js:
```diff
@@ -59,6 +62,7 @@ function WpcomNux() {
 		if ( isWpcomNuxEnabled && ! isSPTOpen ) {
 			recordTracksEvent( 'calypso_editor_wpcom_nux_open', {
 				is_gutenboarding: window.calypsoifyGutenberg?.isGutenboarding,
+				is_manually_opened: isGuideManuallyOpened,
```
```suggestion
				is_manually_opened: !! isGuideManuallyOpened,
```

I think I've noticed the `is_manually_opened` prop can be missing from the event, and I think it's because the value defaults to `undefined`. Forcing the value to a boolean means it'll always send a `false` instead of just a falsey value.

autumnfjeld replied:
Ah, yes.  For the[ Tour, I set the default value in the store to false](https://github.com/Automattic/wp-calypso/blob/10cb217a9e8dc43f72cd70629cd195db062e60b5/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/src/store.js#L15), but I didn't do that for the NUX.    I had been debating refactoring the NUX store to be structured like the store in the Tour component, so I just went ahead and did that for clarity & consistency and[ set the default value for `isGuideManuallyOpened`](https://github.com/Automattic/wp-calypso/blob/4bcb955a8ab4329cba1b555319ac27c2ea003b6c/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js#L15), which solves the issue you have pointed out. #:+1:

p-jackson replied:
Nice, what you've done is probably nicer tbh

p-jackson reviewed (commented):
It's working for me now 😅 I'm sure the code looked different when I reviewed it yesterday, but I see the code that sends the track event for the existing nux modal now. Or maybe the thing below is what threw me off.

When the nux modal is opened automatically, `isGuideManuallyOpened` is `undefined`. That's sorta fine because it's falsey so works out ok in JS. But when sending an `undefined` prop in tracks it just leaves the value out of the event. I think it would be better to explicitly set the value to `false` (which I pointed out below) because maybe it'd be easier to distinguish between the two events in our tools that way?

autumnfjeld committed:
NUX modal: add isTourManuallyOpened to list of useEffect dependencies

autumnfjeld commented:
@roo2 
> Also I wasn't certain the best way to test the older nux modal, it makes more sense to me to simply disable the SHOW_WELCOME_TOUR flag, would't that be more correct then editing the `load_wpcom_block_editor_nux` function? (I'm getting different results)
> 

Many ways to skin a cat.  Your approach is completely valid.  I chose to keep SHOW_WELCOME_TOUR enabled because it has two effects 1) set [nux status as enabled](https://github.com/Automattic/wp-calypso/blob/01d0ee50f39f258e1be7d3b2c165a876e49cf42b/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-welcome-tour/class-wp-rest-wpcom-block-editor-nux-status-controller.php#L60) and 2) [show the new Tour](https://github.com/Automattic/wp-calypso/blob/12f778c15658ed71646145aab9f83586900d04e4/apps/editing-toolkit/editing-toolkit-plugin/full-site-editing-plugin.php#L246).     But disabling SHOW_WELCOME_TOUR is fine, then just need to ensure the nux status is dealt with (again many ways to do that).

autumnfjeld committed:
NUX modal store: small refactor to bring consistency with the Welcome Tour; set default value for isGuideManuallyOpened

autumnfjeld commented:
@roo2 

> For the old nux modal, I tried to test it by disabling the SHOW_WELCOME_TOUR flag in sandbox.php
> 
> I didn't see an is_manually_opened flag being passed along to the tracking endpoint 🔴
> I also tried to show the old nux modal my editing the load_wpcom_block_editor_nux function:
> 
> When I deleted the local storage and reloaded the page, is_manually_opened was not send to the tracking endpoint❓
> ** It should probably send is_manually_opened: false the same as the tour
> I dismissed the modal and reloaded the page, the modal did not show ✅
> I opened the guide through the menu link and an event was fired with is_manually_opened: true
> I reloaded the page and the nux modal was shown again with is_manually_opened:true ❓
> ** I think it should be set to false the second time? or just not open?

Thanks for catching all these!  I addressed two issues that should solve this:
1)  https://github.com/Automattic/wp-calypso/pull/48846/commits/7089ae20e67fa58a124df1e6c5e7178fa2490fb7
2)  https://github.com/Automattic/wp-calypso/pull/48846/commits/4bcb955a8ab4329cba1b555319ac27c2ea003b6c

p-jackson reviewed (approved):
I kept on testing it wrong but got there in the end :D This is working as expected. :shipit:

roo2 commented:
I retested and the older nux modal now reports `is_manually_opened` correctly ✅
